wurst/file/Base64.wurst

```diff
@@ -0,0 +1,284 @@
+import ErrorHandling
+
+/*  Base64 package provides an encoder/decoder pair that allows serializing and deserializing of any data.
+
+    Encode Data iteratively:
+    > let encoder = new Base64Encoder()
+    > encoder.writeByte(25)
+    > encoder.writeInt(100500)
+    > let data = encoder.intoData() // the encoder is consumed, no need to destroy it
+
+    Decode Data iteratively:
+    > let decoder = new Base64Decoder()
+    > while data.hasChunk()
+    >   decoder.append(data.readChunk())
+    > let bytes = decoder.intoData() // the decoder is consumed, no need to destroy it
+    > let byte = bytes.readByte()
+    > let number = bytes.readInt()
+
+    This package also provides convenient methods for ChunkedStrings and ByteBuffers.
+
+    Encode Data:
+    > let data = bytes.encodeBase64()
+
+    Decode Data:
+    > let bytes = data.decodeBase64()
+*/
+
+/**
+    Specifies how many characters to encode per a single `execute()` call.
+    This value has been tuned to work under all optimization settings and
+    with stacktraces included.
+**/
+@configurable public constant ENCODES_PER_ROUND = 1000
+/**
+    Specifies how many chunks to decode per a single `execute()` call.
+    This value has been tuned to work under all optimization settings and
+    with stacktraces included.
+**/
+@configurable public constant DECODES_PER_ROUND = 100
+
+// RFC 4648 compliant Base64 charmap.
+constant CHARMAP = [
+    "A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P",
+    "Q","R","S","T","U","V","W","X","Y","Z","a","b","c","d","e","f",
+    "g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v",
+    "w","x","y","z","0","1","2","3","4","5","6","7","8","9","+","/"
+]
+
+// RFC 4648 compliant Base64 reverse charmap.
+constant REVERSE_CHARMAP = [
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 62, 0, 0, 0, 63, 52, 53, 54,
+    55, 56, 57, 58, 59, 60, 61, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2,
+    3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
+    20, 21, 22, 23, 24, 25, 0, 0, 0, 0, 0, 0, 26, 27, 28, 29, 30,
+    31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
+    48, 49, 50, 51, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+]
+
+public class Base64Encoder
+    private ChunkedString chunkedString
+
+    // We append only one char each time, and strings in WC3 get fully copied on each concatenation.
+    // To optimize it and lower the times of copying quite big chunks, append by smaller chunks.
+    private var stringBuffer = ""
```

Comment:
Isn't this a bit superflous to the buffer inside the ChunkedString?

Reply:
The thing I'm worried about is the following:
Imagine we have the buffer inside the ChunkedString with 160 chars. Here we know we append by exactly one symbol. We append a char 32 times. So the buffer inside the string becomes 161 chars, 162 chars, 163 chars...192 chars long... And copies each time entirely. So 161 chars is copied in the first append, 162 in the second, 163 in the third...192 chars in the end.
If we append to this buffer previously:
1 char get copied, 2 chars get copied...32 chars get copied... This buffer is full. Append to the ChunkedString buffer: 160 + 32 chars get copied one time.
Look at the difference here: 1+2+...32 + 32 + 160 = 720 character copies with this optimization VS 160+161+...200 = 5648 character copies without.